orttraining/orttraining/python/training/ortmodule/_utils.py

```diff
@@ -138,17 +144,28 @@ def check_for_name_collisions(ortmodule: torch.nn.Module, user_module: torch.nn.
             if attribute_name.startswith('__'):
                 continue
 
+            # if the attribute is not a torch attribute, or if the torch attribute
+            # corresponding to attribute_name is not a method or the user attribute
+            # does not equal the torch attribute, then this is a user defined method.
             if attribute_name not in torch_module_attributes or \
                 not inspect.ismethod(torch_module_attributes[attribute_name]) or \
                 attribute.__func__ != torch_module_attributes[attribute_name].__func__:
 
+                # forward is expected to be defined by the user.
                 if attribute_name == 'forward':
                     continue
 
                 # This is a user defined/overriden method. Check for collisions.
                 if attribute_name in ortmodule_attributes:
+                    # This is a user defined method, issue a warning.
                     warnings.warn(f"User Module's attribute name {attribute_name} collides with ORTModule's attribute name. "
                         "User Module's method may not be called upon invocation through ORTModule.")
+                else:
+                    # This is a custom method, copy it and bind the copy to ORTModule.
+                    # This is needed for cases where the user's custom method makes invokes
+                    # the forward method. It should go through ORTModule's forwrad implementation
```

Comment:
This is the problem of PyTorch Lightening that we are trying to address. Here is an analogous problem:
```py
class MyModel(torch.nn.Module):
    def __init__(self):
        ...

    def forward(self, ...):
        ...

    def training_step(self, ...):
        return self(...) # call to self.forward


model = ORTModule(MyModel())
model.training_step() # executes PyTorch module's forward and not ORTModule's forward.
```

As you can see, `training_step` is a custom methods that internally invokes forward. So, if called, it will not go through ```ORTModule```

If you follow [this link](https://pytorch-lightning.readthedocs.io/en/latest/_modules/pytorch_lightning/plugins/training_type/training_type_plugin.html#TrainingTypePlugin) and search for `training_step`, the problem will become evident that the way PyTorch Lightning plugins are defined is that they call the method `training_step` on the original module. And this is the source of the problem.

These models call their own forward from inside their functions. The changes I made makes copies of such methods and binds them to `ORTModule`. So if these methods are called, `ORTModule's` forward implementation is invoked.